src/targets/gpu/include/migraphx/gpu/gemm_impl.hpp

```diff
@@ -24,28 +24,488 @@
+                                           rocblas_int* list,
+                                           rocblas_int* list_size) const
+    {
+        auto common_args = create_gemm_ex_args_common(ctx, args);
+        auto ded_args    = pack(int8_flag, list, list_size);
+        return pack_join(common_args, ded_args);
+    }
+
+    /**
+     * Find best rocBLAS solution:  Get list of solutions and try them all, returning the index
+     * of the fastest one.
+     */
+    int tune(context& ctx, const std::vector<shape>& input_shapes) const
+    {
+        std::vector<argument> input_args;
+        std::transform(input_shapes.begin(),
+                       input_shapes.end(),
+                       std::back_inserter(input_args),
+                       [](const shape& x) { return to_gpu(generate_argument(x)); });
+
+        // Find out how many solutions there are
+        rocblas_int list_size = 0;
+        std::vector<rocblas_int> solution_indices;
+        if(strided_batched)
+        {
+            auto arg_list = create_strided_batched_gemm_get_solutions_args(
+                ctx, input_args, nullptr, &list_size);
+            rocblas_invoke(&rocblas_gemm_strided_batched_ex_get_solutions, arg_list);
+            solution_indices.resize(list_size);
+            auto arg_list_solutions = create_strided_batched_gemm_get_solutions_args(
+                ctx, input_args, solution_indices.data(), &list_size);
+            rocblas_invoke(&rocblas_gemm_strided_batched_ex_get_solutions, arg_list_solutions);
+        }
+        else
+        {
+            auto arg_list = create_gemm_ex_get_solutions_args(ctx, input_args, nullptr, &list_size);
+            rocblas_invoke(&rocblas_gemm_ex_get_solutions, arg_list);
+            solution_indices.resize(list_size);
+            auto arg_list_solutions = create_gemm_ex_get_solutions_args(
+                ctx, input_args, solution_indices.data(), &list_size);
+            rocblas_invoke(&rocblas_gemm_ex_get_solutions, arg_list_solutions);
+        }
+
+        double bestTime = std::numeric_limits<double>::max();
+        // Initialize to default solution index
+        rocblas_int bestSol = 0;
+        for(auto sol : solution_indices)
+        {
+            // Warmup: the first few calls to an op. may not be representative since there is
+            // more time taken initializing caches, etc. so we won't time them.
+            for(rocblas_int cc = 0; cc < cold_calls; ++cc)
+            {
+                run(ctx, input_args, sol);
+            }
+            double host_time = 0.0;
+            // Define the function to be timed
+            auto run_func = [&]() {
+                run(ctx, input_args, sol);
+                ctx.finish();
+            };
+            for(rocblas_int hc = 0; hc < hot_calls; ++hc)
+            {
+                ctx.finish();
+                host_time += time<microseconds>(run_func);
+            }
```

Comment:
This should just use the `time_op` function from the driver. We just need to move it to the gpu lib so it can be reused here.

Reply:
You'll have to specify what you have in mind for that.  gemm_impl doesn't use instances of the Migraphx operation, which is what time_op takes as an argument.  Instead, we're making the rocblas calls directly.  So we can't reuse it directly--what would we substitute here?  (Also, does this change belong in the same PR?)